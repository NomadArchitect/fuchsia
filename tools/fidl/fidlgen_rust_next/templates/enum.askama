{% let name = enm.name.decl_name().camel() %}
{% let natural_int = compiler.natural_int(enm.ty) %}
{% let wire_int = compiler.wire_int(enm.ty) %}

{% if let Some(doc) = enm.attributes.doc_string() %}
    #[doc = "{{ doc.escape_default() }}"]
{% endif %}
#[derive(
    Clone,
    Copy,
    {% if compiler.config.emit_debug_impls -%}
        Debug,
    {%- endif %}
)]
#[repr({{ natural_int }})]
pub enum {{ name }} {
    {% for member in enm.members -%}
        {{ member.name.camel() }} = {{ member.value.value }},
    {%- endfor %}
    {%- if !enm.is_strict -%}
        Unknown({{ natural_int }}),
    {%- endif %}
}

impl ::fidl_next::Encodable for {{ name }} {
    type Encoded<'buf> = Wire{{ name }};
}

impl<___E> ::fidl_next::Encode<___E> for {{ name }}
where
    ___E: ?Sized,
{
    fn encode(
        &mut self,
        _: &mut ___E,
        slot: ::fidl_next::Slot<'_, Self::Encoded<'_>>,
    ) -> Result<(), ::fidl_next::EncodeError> {
        ::fidl_next::munge!(let Wire{{ name }} { mut value } = slot);
        *value = {{ wire_int }}::from(match *self {
            {% for member in enm.members %}
                Self::{{ member.name.camel() }} => {{ member.value.value }},
            {% endfor %}
            {% if !enm.is_strict %}
                Self::Unknown(value) => value,
            {% endif %}
        });

        Ok(())
    }
}

impl ::core::convert::From<Wire{{ name }}> for {{ name }} {
    fn from(wire: Wire{{ name }}) -> Self {
        match {{ natural_int }}::from(wire.value) {
            {% for member in enm.members %}
                {{ member.value.value }} => Self::{{ member.name.camel() }},
            {% endfor %}
            {% if enm.is_strict %}
                _ => unsafe { ::core::hint::unreachable_unchecked() },
            {% else %}
                value => Self::Unknown(value),
            {% endif %}
        }
    }
}

impl ::fidl_next::TakeFrom<Wire{{ name }}> for {{ name }} {
    fn take_from(from: &mut Wire{{ name }}) -> Self {
        Self::from(*from)
    }
}

/// The wire type corresponding to [`{{ name }}`].
#[derive(
    Clone,
    Copy,
    {% if compiler.config.emit_debug_impls -%}
        Debug,
    {%- endif %}
)]
#[repr(transparent)]
pub struct Wire{{ name }} {
    value: {{ wire_int }},
}

impl Wire{{ name }} {
    {% for member in enm.members %}
        pub const {{ member.name.screaming_snake() }}: Wire{{ name }} = Wire{{ name }} {
            value:
            {% match enm.ty %}
                {% when IntType::Int8 %} {{ member.value.value }}
                {% when IntType::Uint8 %} {{ member.value.value }}
                {% else %} {{ wire_int }}::from_native({{ member.value.value }})
            {% endmatch %}
        };
    {% endfor %}
}

unsafe impl<___D> ::fidl_next::Decode<___D> for Wire{{ name }}
where
    ___D: ?Sized,
{
    fn decode(
        slot: ::fidl_next::Slot<'_, Self>,
        _ : &mut ___D,
    ) -> Result<(), ::fidl_next::DecodeError> {
        {% if enm.is_strict %}
            ::fidl_next::munge!(let Self { value } = slot);

            match {{ natural_int }}::from(*value) {
                {% for member in enm.members %}
                    | {{ member.value.value }}
                {% endfor %}
                => (),
                unknown => return Err(::fidl_next::DecodeError::InvalidEnumOrdinal(
                    unknown as i128,
                )),
            }
        {% endif %}

        Ok(())
    }
}

impl ::core::convert::From<{{ name }}> for Wire{{ name }} {
    fn from(natural: {{ name }}) -> Self {
        match natural {
            {% for member in enm.members %}
                {{ name }}::{{ member.name.camel() }} => Wire{{ name }}::{{ member.name.screaming_snake() }},
            {% endfor %}
            {% if !enm.is_strict %}
                {{ name }}::Unknown(value) => Wire{{ name }} {
                    value: {{ wire_int }}::from(value),
                }
            {% endif %}
        }
    }
}
