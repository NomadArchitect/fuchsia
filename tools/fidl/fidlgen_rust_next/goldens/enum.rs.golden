// DO NOT EDIT: This file is machine-generated by fidlgen
#![warn(clippy::all)]
#![allow(
    unused_parens,
    unused_variables,
    unused_mut,
    unused_imports,
    unreachable_code,
    nonstandard_style
)]

#[derive(Clone, Copy, Debug)]
#[repr(u32)]
pub enum MyEmptyFlexibleEnum {
    Unknown(u32),
}

impl ::fidl::Encodable for MyEmptyFlexibleEnum {
    type Encoded<'buf> = WireMyEmptyFlexibleEnum;
}

impl<___E> ::fidl::Encode<___E> for MyEmptyFlexibleEnum
where
    ___E: ?Sized,
{
    fn encode(
        &mut self,
        _: &mut ___E,
        slot: ::fidl::Slot<'_, Self::Encoded<'_>>,
    ) -> Result<(), ::fidl::EncodeError> {
        ::fidl::munge!(let WireMyEmptyFlexibleEnum { mut value } = slot);
        *value = ::fidl::u32_le::from(match *self {
            MyEmptyFlexibleEnum::Unknown(value) => value,
        });

        Ok(())
    }
}

impl ::core::convert::From<WireMyEmptyFlexibleEnum> for MyEmptyFlexibleEnum {
    fn from(wire: WireMyEmptyFlexibleEnum) -> Self {
        match u32::from(wire.value) {
            value => MyEmptyFlexibleEnum::Unknown(value),
        }
    }
}

impl ::fidl::TakeFrom<WireMyEmptyFlexibleEnum> for MyEmptyFlexibleEnum {
    fn take_from(from: &mut WireMyEmptyFlexibleEnum) -> Self {
        MyEmptyFlexibleEnum::from(*from)
    }
}

#[derive(Clone, Copy, Debug)]
#[repr(transparent)]
pub struct WireMyEmptyFlexibleEnum {
    value: ::fidl::u32_le,
}

impl WireMyEmptyFlexibleEnum {}

unsafe impl<___D> ::fidl::Decode<___D> for WireMyEmptyFlexibleEnum
where
    ___D: ?Sized,
{
    fn decode(slot: ::fidl::Slot<'_, Self>, _: &mut ___D) -> Result<(), ::fidl::DecodeError> {
        Ok(())
    }
}

impl ::core::convert::From<MyEmptyFlexibleEnum> for WireMyEmptyFlexibleEnum {
    fn from(natural: MyEmptyFlexibleEnum) -> Self {
        match natural {
            MyEmptyFlexibleEnum::Unknown(value) => {
                WireMyEmptyFlexibleEnum { value: ::fidl::u32_le::from(value) }
            }
        }
    }
}

#[derive(Clone, Copy, Debug)]
#[repr(u32)]
pub enum MyFlexibleEnum {
    Foo = 1,
    Bar = 2,
    Unknown(u32),
}

impl ::fidl::Encodable for MyFlexibleEnum {
    type Encoded<'buf> = WireMyFlexibleEnum;
}

impl<___E> ::fidl::Encode<___E> for MyFlexibleEnum
where
    ___E: ?Sized,
{
    fn encode(
        &mut self,
        _: &mut ___E,
        slot: ::fidl::Slot<'_, Self::Encoded<'_>>,
    ) -> Result<(), ::fidl::EncodeError> {
        ::fidl::munge!(let WireMyFlexibleEnum { mut value } = slot);
        *value = ::fidl::u32_le::from(match *self {
            MyFlexibleEnum::Foo => 1,
            MyFlexibleEnum::Bar => 2,
            MyFlexibleEnum::Unknown(value) => value,
        });

        Ok(())
    }
}

impl ::core::convert::From<WireMyFlexibleEnum> for MyFlexibleEnum {
    fn from(wire: WireMyFlexibleEnum) -> Self {
        match u32::from(wire.value) {
            1 => MyFlexibleEnum::Foo,
            2 => MyFlexibleEnum::Bar,
            value => MyFlexibleEnum::Unknown(value),
        }
    }
}

impl ::fidl::TakeFrom<WireMyFlexibleEnum> for MyFlexibleEnum {
    fn take_from(from: &mut WireMyFlexibleEnum) -> Self {
        MyFlexibleEnum::from(*from)
    }
}

#[derive(Clone, Copy, Debug)]
#[repr(transparent)]
pub struct WireMyFlexibleEnum {
    value: ::fidl::u32_le,
}

impl WireMyFlexibleEnum {
    pub const FOO: WireMyFlexibleEnum =
        WireMyFlexibleEnum { value: ::fidl::u32_le::from_native(1) };
    pub const BAR: WireMyFlexibleEnum =
        WireMyFlexibleEnum { value: ::fidl::u32_le::from_native(2) };
}

unsafe impl<___D> ::fidl::Decode<___D> for WireMyFlexibleEnum
where
    ___D: ?Sized,
{
    fn decode(slot: ::fidl::Slot<'_, Self>, _: &mut ___D) -> Result<(), ::fidl::DecodeError> {
        Ok(())
    }
}

impl ::core::convert::From<MyFlexibleEnum> for WireMyFlexibleEnum {
    fn from(natural: MyFlexibleEnum) -> Self {
        match natural {
            MyFlexibleEnum::Foo => WireMyFlexibleEnum::FOO,
            MyFlexibleEnum::Bar => WireMyFlexibleEnum::BAR,
            MyFlexibleEnum::Unknown(value) => {
                WireMyFlexibleEnum { value: ::fidl::u32_le::from(value) }
            }
        }
    }
}

#[derive(Clone, Copy, Debug)]
#[repr(u32)]
pub enum MyFlexibleEnumWithCustomUnknown {
    Foo = 1,
    Bar = 2,
    CustomUnknown = 3,
    Unknown(u32),
}

impl ::fidl::Encodable for MyFlexibleEnumWithCustomUnknown {
    type Encoded<'buf> = WireMyFlexibleEnumWithCustomUnknown;
}

impl<___E> ::fidl::Encode<___E> for MyFlexibleEnumWithCustomUnknown
where
    ___E: ?Sized,
{
    fn encode(
        &mut self,
        _: &mut ___E,
        slot: ::fidl::Slot<'_, Self::Encoded<'_>>,
    ) -> Result<(), ::fidl::EncodeError> {
        ::fidl::munge!(let WireMyFlexibleEnumWithCustomUnknown { mut value } = slot);
        *value = ::fidl::u32_le::from(match *self {
            MyFlexibleEnumWithCustomUnknown::Foo => 1,
            MyFlexibleEnumWithCustomUnknown::Bar => 2,
            MyFlexibleEnumWithCustomUnknown::CustomUnknown => 3,
            MyFlexibleEnumWithCustomUnknown::Unknown(value) => value,
        });

        Ok(())
    }
}

impl ::core::convert::From<WireMyFlexibleEnumWithCustomUnknown>
    for MyFlexibleEnumWithCustomUnknown
{
    fn from(wire: WireMyFlexibleEnumWithCustomUnknown) -> Self {
        match u32::from(wire.value) {
            1 => MyFlexibleEnumWithCustomUnknown::Foo,
            2 => MyFlexibleEnumWithCustomUnknown::Bar,
            3 => MyFlexibleEnumWithCustomUnknown::CustomUnknown,
            value => MyFlexibleEnumWithCustomUnknown::Unknown(value),
        }
    }
}

impl ::fidl::TakeFrom<WireMyFlexibleEnumWithCustomUnknown> for MyFlexibleEnumWithCustomUnknown {
    fn take_from(from: &mut WireMyFlexibleEnumWithCustomUnknown) -> Self {
        MyFlexibleEnumWithCustomUnknown::from(*from)
    }
}

#[derive(Clone, Copy, Debug)]
#[repr(transparent)]
pub struct WireMyFlexibleEnumWithCustomUnknown {
    value: ::fidl::u32_le,
}

impl WireMyFlexibleEnumWithCustomUnknown {
    pub const FOO: WireMyFlexibleEnumWithCustomUnknown =
        WireMyFlexibleEnumWithCustomUnknown { value: ::fidl::u32_le::from_native(1) };
    pub const BAR: WireMyFlexibleEnumWithCustomUnknown =
        WireMyFlexibleEnumWithCustomUnknown { value: ::fidl::u32_le::from_native(2) };
    pub const CUSTOM_UNKNOWN: WireMyFlexibleEnumWithCustomUnknown =
        WireMyFlexibleEnumWithCustomUnknown { value: ::fidl::u32_le::from_native(3) };
}

unsafe impl<___D> ::fidl::Decode<___D> for WireMyFlexibleEnumWithCustomUnknown
where
    ___D: ?Sized,
{
    fn decode(slot: ::fidl::Slot<'_, Self>, _: &mut ___D) -> Result<(), ::fidl::DecodeError> {
        Ok(())
    }
}

impl ::core::convert::From<MyFlexibleEnumWithCustomUnknown>
    for WireMyFlexibleEnumWithCustomUnknown
{
    fn from(natural: MyFlexibleEnumWithCustomUnknown) -> Self {
        match natural {
            MyFlexibleEnumWithCustomUnknown::Foo => WireMyFlexibleEnumWithCustomUnknown::FOO,
            MyFlexibleEnumWithCustomUnknown::Bar => WireMyFlexibleEnumWithCustomUnknown::BAR,
            MyFlexibleEnumWithCustomUnknown::CustomUnknown => {
                WireMyFlexibleEnumWithCustomUnknown::CUSTOM_UNKNOWN
            }
            MyFlexibleEnumWithCustomUnknown::Unknown(value) => {
                WireMyFlexibleEnumWithCustomUnknown { value: ::fidl::u32_le::from(value) }
            }
        }
    }
}

#[derive(Clone, Copy, Debug)]
#[repr(u32)]
pub enum MyStrictEnum {
    Foo = 1,
    Bar = 2,
}

impl ::fidl::Encodable for MyStrictEnum {
    type Encoded<'buf> = WireMyStrictEnum;
}

impl<___E> ::fidl::Encode<___E> for MyStrictEnum
where
    ___E: ?Sized,
{
    fn encode(
        &mut self,
        _: &mut ___E,
        slot: ::fidl::Slot<'_, Self::Encoded<'_>>,
    ) -> Result<(), ::fidl::EncodeError> {
        ::fidl::munge!(let WireMyStrictEnum { mut value } = slot);
        *value = ::fidl::u32_le::from(match *self {
            MyStrictEnum::Foo => 1,
            MyStrictEnum::Bar => 2,
        });

        Ok(())
    }
}

impl ::core::convert::From<WireMyStrictEnum> for MyStrictEnum {
    fn from(wire: WireMyStrictEnum) -> Self {
        match u32::from(wire.value) {
            1 => MyStrictEnum::Foo,
            2 => MyStrictEnum::Bar,
            _ => unsafe { ::core::hint::unreachable_unchecked() },
        }
    }
}

impl ::fidl::TakeFrom<WireMyStrictEnum> for MyStrictEnum {
    fn take_from(from: &mut WireMyStrictEnum) -> Self {
        MyStrictEnum::from(*from)
    }
}

#[derive(Clone, Copy, Debug)]
#[repr(transparent)]
pub struct WireMyStrictEnum {
    value: ::fidl::u32_le,
}

impl WireMyStrictEnum {
    pub const FOO: WireMyStrictEnum = WireMyStrictEnum { value: ::fidl::u32_le::from_native(1) };
    pub const BAR: WireMyStrictEnum = WireMyStrictEnum { value: ::fidl::u32_le::from_native(2) };
}

unsafe impl<___D> ::fidl::Decode<___D> for WireMyStrictEnum
where
    ___D: ?Sized,
{
    fn decode(slot: ::fidl::Slot<'_, Self>, _: &mut ___D) -> Result<(), ::fidl::DecodeError> {
        ::fidl::munge!(let Self { value } = slot);

        match u32::from(*value) {
            1 | 2 => (),
            unknown => return Err(::fidl::DecodeError::InvalidEnumOrdinal(unknown as usize)),
        }

        Ok(())
    }
}

impl ::core::convert::From<MyStrictEnum> for WireMyStrictEnum {
    fn from(natural: MyStrictEnum) -> Self {
        match natural {
            MyStrictEnum::Foo => WireMyStrictEnum::FOO,
            MyStrictEnum::Bar => WireMyStrictEnum::BAR,
        }
    }
}
